File: programmers/소수 찾기 Level - 2/solution.py
def DFS(numbers, nums, stack, visit, isPrime):
    if stack:
        nums.add(int("".join(stack)))
            
    for i in range(len(numbers)):
        if not visit[i]:
            visit[i] = True
            stack.append(numbers[i])
            DFS(numbers, nums, stack, visit, isPrime)
            visit[i] = False
            stack.pop()

    
def solution(numbers):
    answer = 0
    
    isPrime = [False] + [False] + [True] * (9999997)
    
    for i in range(2, 9999997):
        if isPrime[i]:
            for j in range(i * 2, 9999997, i):
                isPrime[j] = False
        
    stack = []
    nums = set()
    visit = [False] * len(numbers)
    
    DFS(numbers, nums, stack, visit, isPrime)
    
    for num in nums:
        if isPrime[num]:
            answer += 1
            
    return answer

File: programmers/소수 찾기 Level - 2/Solution.java
import java.util.*;

class Solution {
    private static boolean[] visit;
    private static Stack<Integer> stack = new Stack<>();
    private static boolean[] isPrimeNumber = new boolean[10000001];
    private static Set<Integer> candidates = new HashSet<>();
    
    private static void primeNumberSetting() {
        isPrimeNumber[0] = isPrimeNumber[1] = true;
        
         for (int i = 2; i * i < isPrimeNumber.length; i++){
            if (!isPrimeNumber[i]) {
            	for (int j = i * i; j < isPrimeNumber.length; j+=i) {
                    isPrimeNumber[j] = true;
                }
            }        
        }
    }
    
    private static void dfs(String numbers) {
        if (!stack.isEmpty()) {
            StringBuilder sb = new StringBuilder();
            for (int number : stack) {
                sb.append(number);    
            }
            
           candidates.add(Integer.parseInt(sb.toString()));
        }
        
        for (int i = 0; i < numbers.length(); i++) {
            if (!visit[i]) {
                visit[i] = true;
                stack.add(numbers.charAt(i) - '0');
                dfs(numbers);
                visit[i] = false;
                stack.pop();
            }
        }
    }
    
    public int solution(String numbers) {
        int answer = 0;
        visit = new boolean[numbers.length()];
        primeNumberSetting();
        dfs(numbers);
        
        for (int candidate : candidates) {
            if (!isPrimeNumber[candidate]) {
                answer++;
            }
        }
        
        return answer;
    }
}